### PR TITLE
Feat CI: multiple platforms & coverage reports

### PR DESCRIPTION
**Pull request by sphamba**
# Prework

- [x] I understand and agree to this repository's [code of conduct](https://github.com/ltelab/disdrodb/blob/main/CODE_OF_CONDUCT.md).
- [x] I understand and agree to this repository's [contributing guidelines](https://github.com/ltelab/disdrodb/blob/main/CONTRIBUTING.md).
- [x] I have already submitted an [issue](https://github.com/ltelab/disdrodb/issues) or [discussion thread](https://github.com/ltelab/disdrodb/discussions) to discuss my idea with the maintainers.

**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [x] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and communicate accordingly:

**The PR fulfills these requirements:**

- [x] It's submitted to the branch named as follow : 
  - Add a reader: `reader-<institute>-<campaign>`
  - Fix a bug: `bugfix-<some_key>-<word>`
  - Improve the doc: `doc-<some_key>-<word>`
  - Add a new feature: `feature-<some_key>-<word>`
  - Refactor some code: `refactor-<some_key>-<word>`
  - Optimize some code: `optimize-<some_key>-<word>`
- [ ] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)
- [ ] Don't forget to link PR to issue if you are solving one.
- [x] All tests are passing.
- [ ] New/updated tests are included

If adding a **new feature**, the PR's description includes:

- [x] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

**Other information:**

# Related GitHub issues and pull requests

*none*

# Summary

## Main changes
- Add CI tests on Python 3.8 to 3.10, plus 3.10 with latest dependencies (pinned dependencies are specified in `requirements.txt` and without version in `pyproject.toml`)
- Add Coveralls and Codecov reports in CI

## Notes
- It seems to be common practice to have the tests and coverage checks run at each push. This can be changed in [tests.yml:2](https://github.com/EPFL-ENAC/LTE-disdrodb/blob/98513b1dc58ab6bd87b7ef51eb6d1c16850a5d54/.github/workflows/tests.yml#L2)
- To have the tests run also on Windows and macOS, change [tests.yml:11](https://github.com/EPFL-ENAC/LTE-disdrodb/blob/98513b1dc58ab6bd87b7ef51eb6d1c16850a5d54/.github/workflows/tests.yml#L11) to
`os: ["macos-latest", "ubuntu-latest", "windows-latest"]`
Note that compared to Ubuntu, the counted CPU minutes are [multilied by 2 on Windows and by 10 on macOS](https://docs.github.com/en/billing/managing-billing-for-github-actions/about-billing-for-github-actions#minute-multipliers), so I did not set them
- To have the Coveralls report integrated to the PRs, you must:
  - [x] Invite the @coveralls user to the repository as a collaborator
  - [ ] Enable the "LEAVE COMMENTS?" setting in the "Pull Request Alerts" area in the Repo Settings page inside the Coveralls app
- For now, `requirements.txt` is kept for installing dependencies in CI. Please tell us where you would want to have micromamba used (in CI or for contributors) and how it could be advantageous? I think that `requirements.txt` should be kept anyway for contributors using `pip`

## Other code quality check tools
- CodeFactory.io
  - code quality grades (A-F)
  - code issues: complexity, style, mintainability, performance...
  - integration in Github, Slack
- CodeBeat.io: similar but less insight. Nothing much shown on website
- CodeScene, Codacy, Code Climate: all paid

### CODE REVIEW

**Comment by ghiggi** [comment]
Thanks for this PR. Answering your questions: 
1. Very good point to run tests  and coverage checks on each push
2 Let's keep using pip to install the dependencies in the CI 
3. Let's continue to run the CI only on Linux for the moment (and then I will discuss with Alexis what he aims to do)
4. I already set up CodeScene and Codebeat and should pop up when committing new code. (https://codebeat.co/projects/github-com-ltelab-disdrodb-main/pull_requests) 
5. I also set up the experimental Copilot for Pull Requests for this repo :) 
6. I invited the @coveralls user to the repository as a collaborator but I was not able to find the `"LEAVE COMMENTS?" setting in the "Pull Request Alerts" area in the Repo Settings page inside the Coveralls app`.  I added @sphamba to the maintainers of the repo so that he can also search for it. Btw coverall is already seeing this PR: https://coveralls.io/builds/58298909

**Comment by ghiggi** [comment]
copilot:summary

**Comment by sphamba** [comment]
For 5., as the owner of the repo, you must log into coveralls.io. There, you can see your currently enabled repositories, access their settings, and have the "LEAVE COMMENTS?" option under the "PULL REQUESTS ALERTS" section.

**Comment by regislon** [comment]
Hello @ghiggi could you please correct or complete this following table : 


| Name | Aim | pre-commit | Github action | Link | Python version | OS |
| --- | --- | --- | --- | --- | --- | --- |
| pytest | Execute unit tests and functional tests | 👍 | 👍 |  |  |  |
| Black  | code format checher to ensure a consistent style. | 👍 | 👍 |  | No python version (Black version 22.8.0) | ubuntu-latest |
| Ruff  | Python linter  | 👍 | should be added to this PR |  | ? | ubuntu-latest |
| Coverage | Measure the code coverage of the project's unit tests | to test | 👍 |  | all versions according to github workflow | ubuntu-latest |
| Codecov | Uses the "coverage" package to generate a code coverage report. | 👎 | 👍 |  | all versions according to github workflow | ubuntu-latest |
| Coveralls | Uses the "coverage" to track the quality of your code over time. | 👎 | 👍 | https://coveralls.io/builds/58298909 | all versions according to github workflow | ubuntu-latest |
| CodeBeat | Automated code review and analysis tools | 👎 | 👎 | https://codebeat.co/projects/github-com-ltelab-disdrodb-main/security_issues | ? | ? |
| CodeScene | Automated code review and analysis tools | 👎 | 👎 | ? | ? | ? |


Will be used for the doc as well.

Thank you

**Comment by ghiggi** [comment]
@regislon I added some stuffs ... but maybe it's worth so separate what is a pre-commit and a CI/CD tools run with GitHub Actions...

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+          - os: "ubuntu-latest"
+            python-version: "3.10"
+            experimental: false
+            coverage: true
+
+    steps:
+      - uses: actions/checkout@v3
+
+      - name: Set up Python
```
Comment (ghiggi) [inline .github/workflows/tests.yml:27]:
```suggestion
      - name: Set up Python ${{ matrix.python-version }}

```

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+            pip install -r requirements.txt
+            pip install -e .
+          fi
+        shell: bash
+
+      - name: Test with pytest
+        run: |
+          coverage run -m pytest
+
+      - name: Codecov
```
Comment (ghiggi) [inline .github/workflows/tests.yml:48]:
```suggestion
      - name: CodeCov

```

File: .github/workflows/tests.yml [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+          if echo ${{ matrix.experimental }} | grep -c "true"
+          then
+            pip install -e .[dev]
+          else
+            pip install -r requirements.txt
+            pip install -e .
+          fi
+        shell: bash
+
+      - name: Test with pytest
```
Comment (ghiggi) [inline .github/workflows/tests.yml:44]: This also run the unit tests or just check the coverage?

**Comment by ghiggi** [comment]
For the precommit config yaml you can take inspiration from this one: https://github.com/pangeo-data/zarr-proxy/blob/main/.pre-commit-config.yaml

File: .github/workflows/tests.yml (hunk 3 above)
Comment (sphamba) [reply .github/workflows/tests.yml:44]: Tests are also runned and failure is dectected as usual

**Comment by regislon** [comment]
@ghiggi, I agree that CI must be separated from the precommit process. Precommit modifications will be done in another PR tomorrow.

Regarding the documentation, I think it would be nice to include the above table in the following sections:
* ["Contributing process" section ](https://disdrodb.readthedocs.io/en/latest/contributors_guidelines.html#push-your-changes-to-your-fork-repository) : Include the table without the external link to inform contributors how the CI will check their code.
* [Maintainers guidklines](https://disdrodb.readthedocs.io/en/latest/maintainers_guidelines.html#testing-processes) section : Include information on the CI process and links, but exclude the pre-commit part.

What do you think?

To add Ruff to the CI, I propose the following steps:

1. Validate this PR.
2. Work on the pre-commit PR to add Ruff. There may be corrections needed in the code.
3. Create another PR to add Ruff to the CI.

Do you agree, @ghiggi? Is there anything still bocking the PR ? can you accept it @ghiggi

**Comment by ghiggi** [comment]
@regislon 100 % agree with the next steps

**Pull request merged**